src/lib.rs

```diff
@@ -1,6 +1,343 @@
+fn get_pcp_root() -> PathBuf {
+    match env::var_os("PCP_DIR") {
+        Some(val) => PathBuf::from(val),
+        None => PathBuf::from(MAIN_SEPARATOR.to_string())
+    }
+}
+
+fn init_pcp_conf(pcp_root: &PathBuf) -> io::Result<()> {
+    // load variables from pcp_root/etc/pcp.conf into environment
+    let mut default_conf = pcp_root.clone();
+    default_conf.push("etc");
+    default_conf.push("pcp.conf");
+    parse_pcp_conf(default_conf)?;
+
+    // load variables from pcp_root/$PCP_CONF into environment
+    let mut pcp_conf = pcp_root.clone();
+    pcp_conf.push(env::var_os("PCP_CONF").unwrap_or(OsString::new()));
+    parse_pcp_conf(pcp_conf)
+}
+
+fn parse_pcp_conf(conf_path: PathBuf) -> io::Result<()> {
+    let pcp_conf = File::open(conf_path)?;
+    let buf_reader = BufReader::new(pcp_conf);
+
+    /* According to man 5 pcp.conf, syntax rules for pcp.conf are
+        1. general syntax is PCP_VARIABLE_NAME=value to end of line 
+        2. blank lines and lines begining with # are ignored
+        3. variable names that aren't prefixed with PCP_ are silently ignored
+        4. there should be no space between the variable name and the literal =
+        5. values may contain spaces and should not be quoted
+    */
+    lazy_static! {  
+        static ref RE: Regex =
+            Regex::new("^(PCP_[[:alnum:]_]+)=([^\"\'].*[^\"\']$)").unwrap();
+    }
+
+    for line in buf_reader.lines() {
+        match RE.captures(&line?) {
+            Some(caps) => env::set_var(
+                    caps.get(1).map_or("", |m| m.as_str()),
+                    caps.get(2).map_or("", |m| m.as_str()),),
+            _ => {}
+        }
+    }
+
+    Ok(())
+}
+
+fn get_mmv_dir() -> io::Result<PathBuf> {
+    let pcp_root = get_pcp_root();
+    let mut mmv_dir = pcp_root.clone();
+
+    mmv_dir.push(match env::var_os(PCP_TMP_DIR_KEY) {
+        Some(val) => PathBuf::from(val),
+        None => {
+
+            init_pcp_conf(&pcp_root).ok();
+
+            /* re-check if PCP_TMP_DIR is set after parsing (any) conf files
+               if not, default to OS-specific temp dir and set PCP_TMP_DIR
+               so we don't enter this block again */
+            match env::var_os(PCP_TMP_DIR_KEY) {
+                Some(val) => PathBuf::from(val),
+                None => {
+                    let os_tmp_dir = env::temp_dir();
+                    env::set_var(PCP_TMP_DIR_KEY, os_tmp_dir.to_str().unwrap());
+                    os_tmp_dir
+                }
+            }
+        }
+    });
+
+    mmv_dir.push(MMV_DIR_SUFFIX);
+    fs::create_dir_all(&mmv_dir)?;
+
+    Ok(mmv_dir)
+}
+
+bitflags! {
+    /// Flags used to modify how a client exports metrics
+    pub struct MMVFlags: u32 {
+        /// Metric names aren't prefixed with MMV filename
+        const NOPREFIX = 1;
+        /// PID check is needed
+        const PROCESS  = 2;
+        /// Allow "no value available" values
+        const SENTINEL = 4;
+    }
+}
+
+struct MMVWriterInfo {
```

Comment:
When we implement metrics/indoms, we'll need to pass around some counts and section offsets as well. And since tuple attributes are unnamed, it'll be difficult to keep track of which value is what.